internal/web/server.go

```diff
@@ -38,60 +47,102 @@ type ServerConfigSupport struct {
 	BootstrapCheckError       error
 }
 
-func Start(ctx context.Context, support *ServerConfigSupport) error {
-	pkgTemplate, err := template.ParseFS(embededFs, "templates/packages.html")
+func (h *Hub) ws(w http.ResponseWriter, r *http.Request) {
+	wsClient, err := NewWsClient(h, w, r)
 	if err != nil {
-		return err
+		fmt.Fprintf(os.Stderr, "Failed to create WebSocket client: %v", err)
+		return
 	}
-	supportTemplate, err := template.ParseFS(embededFs, "templates/support.html")
-	if err != nil {
-		return err
+	h.Register <- wsClient
+	go wsClient.HandleReads()
+	go wsClient.HandleWrites()
+}
+
+func init() {
+	loadTemplates()
+}
+
+func loadTemplates() {
+	templateFuncs := template.FuncMap{
+		"ToInstallButtonInput": components.ToInstallButtonInput,
 	}
+	baseTemplate = template.Must(
+		template.New("base.html").Funcs(templateFuncs).ParseFS(embededFs, "templates/layout/base.html"))
+	pkgsPageTmpl = template.Must(template.Must(baseTemplate.Clone()).
+		ParseFS(embededFs, "templates/pages/packages.html", "templates/components/*.html"))
+	supportPageTmpl = template.Must(template.Must(baseTemplate.Clone()).
+		ParseFS(embededFs, "templates/pages/support.html", "templates/components/*.html"))
+	installBtnTmpl = template.Must(template.Must(baseTemplate.Clone()).
+		ParseFS(embededFs, "templates/components/installbutton.html"))
+}
 
+func Start(ctx context.Context, support *ServerConfigSupport) error {
 	root, err := fs.Sub(embededFs, "root")
 	if err != nil {
 		return err
 	}
+
+	wsHub := NewHub()
+
 	fileServer := http.FileServer(http.FS(root))
 	http.Handle("/static/", fileServer)
 	http.Handle("/favicon.ico", fileServer)
-	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		if support != nil {
-			err := supportTemplate.Execute(w, support)
+	http.HandleFunc("/ws", wsHub.ws)
+	http.HandleFunc("/install", func(w http.ResponseWriter, r *http.Request) {
+		pkgClient := client.FromContext(ctx)
+		pkgName := r.FormValue("packageName")
+		go func() {
+			status, err := install.NewInstaller(pkgClient).
+				WithStatusWriter(statuswriter.Stderr()).
+				InstallBlocking(ctx, pkgName)
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "An error occurred rendering the response: \n%v\n", err)
+				fmt.Fprintf(os.Stderr, "An error occurred installing %v: \n%v\n", pkgName, err)
 			}
+
+			// broadcast the status update to all clients
+			var bf bytes.Buffer
+			components.RenderInstallButton(&bf, installBtnTmpl, pkgName, status)
+			wsHub.Broadcast <- bf.Bytes()
```

Comment:
A very similar part occurs three times.
I think it would be OK to refactor this in the future.

Reply:
Yes, I think also that the handlers need to be pulled out of there and moved to separate files maybe - I think that could be refactored when implementing the package detail page.